crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -350,3 +373,135 @@ void RSASSA_PSS_PARAMS_free(RSASSA_PSS_PARAMS *params) {
+static int pss_hash_create(const EVP_MD *sigmd, RSA_ALGOR_IDENTIFIER **out) {
+  if (sigmd == NULL) {
+    *out = NULL;
+    return 1;
+  }
+  return pss_parse_nid(EVP_MD_type(sigmd), out);
+}
+
+// pss_mga_create return one on success and zero on failure.
+// When success and the given algorithm is not default (sha1), *out will hold
+// the allocated RSA_ALGOR_IDENTIFIER.
+static int pss_mga_create(const EVP_MD *mgf1md, RSA_MGA_IDENTIFIER **out) {
+  if (mgf1md == NULL || EVP_MD_type(mgf1md) == NID_sha1) {
+    *out = NULL;
+    return 1;
+  }
+  RSA_MGA_IDENTIFIER *mga = RSA_MGA_IDENTIFIER_new();
+  if (mga == NULL) {
+    return 0;
+  }
+  if (pss_parse_nid(EVP_MD_type(mgf1md), &(mga->one_way_hash))) {
+    *out = mga;
+    return 1;
+  }
+  RSA_MGA_IDENTIFIER_free(mga);
+  return 0;
+}
+
+// pss_saltlen_create return one on success and zero on failure.
+// When success and the given len is not default (20), |*out| will hold
+// the allocated RSA_INTEGER.
+static int pss_saltlen_create(int saltlen, RSA_INTEGER **out) {
+  if (saltlen <= 0) {
+    return 0;
+  }
+  if (saltlen == 20) {
+    return 1;
+  }
+  *out = RSA_INTEGER_new();
+  if (*out != NULL) {
+    (*out)->value = saltlen;
+    return 1;
+  }
+  return 0;
+}
+
+int RSASSA_PSS_PARAMS_create(const EVP_MD *sigmd, const EVP_MD *mgf1md,
+                             int saltlen, RSASSA_PSS_PARAMS **out) {
+  // If all parameters are default values don't set pss.
+  if (sigmd == NULL && mgf1md == NULL && saltlen == -2) {
+    return 1;
+  }
+  RSASSA_PSS_PARAMS *pss = RSASSA_PSS_PARAMS_new();
+  if (pss == NULL) {
+    return 0;
+  }
+  if (!pss_hash_create(sigmd, &pss->hash_algor) ||
+      !pss_mga_create(mgf1md, &pss->mask_gen_algor) ||
+      !pss_saltlen_create(saltlen, &pss->salt_len)) {
+    RSASSA_PSS_PARAMS_free(pss);
+    return 0;
+  }
+  *out = pss;
+  return 1;
+}
+
+// nid_to_EVP_MD maps |nid| to the corresponding |EVP_md()| supported by pss.
+// It returns NULL if the |nid| is not matched or supported.
+static const EVP_MD *nid_to_EVP_MD(const int nid) {
+  switch (nid) {
+    case NID_sha1:
+      return EVP_sha1();
+    case NID_sha224:
+      return EVP_sha224();
+    case NID_sha256:
+      return EVP_sha256();
+    case NID_sha384:
+      return EVP_sha384();
+    case NID_sha512:
+      return EVP_sha512();
+    default:
+      OPENSSL_PUT_ERROR(RSA, EVP_R_UNSUPPORTED_ALGORITHM);
+      return NULL;
+  }
+}
+
+// hash_algor_to_EVP_MD return one on success and zero on failure.
+// When success, |*md| will be assigned with the corresponding EVP_md().
+static int hash_algor_to_EVP_MD(RSA_ALGOR_IDENTIFIER *hash_algor,
+                                const EVP_MD **md) {
+  if (hash_algor) {
+    *md = nid_to_EVP_MD(hash_algor->nid);
+  } else {
+    *md = EVP_sha1();
+  }
+  return *md != NULL;
+}
+
+int RSASSA_PSS_PARAMS_get(const RSASSA_PSS_PARAMS *pss, const EVP_MD **md,
+                      const EVP_MD **mgf1md, int *saltlen) {
+  if (pss == NULL) {
```

Comment:
No, 
* the API doc does not say it guarantee that.
* unlikely (less error prone) when using this API -- `mfg1md` and `md` are pointer to pointer.

```c
EVP_MD *md = NULL;
EVP_MD *mgf1md = NULL;
RSASSA_PSS_PARAMS_get(..., &md, &mgf1md, ...)
```

Reply:
I added the NULL check.  
However, `fraudulent input` cannot happen here because this is internal API. Besides, NULL checks are part of method contracts, which may or may not provide the checks.

Reply:
Bad choice of words by me: Any input that goes into our public functions, especially the ones that can be adversarially controlled.